### PR TITLE
Bael 5265: OpenFeign and Spring Exception handler

### PR DESCRIPTION
**Pull request by saikatcse03**
https://jira.baeldung.com/browse/BAEL-5265

### CODE REVIEW

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClient.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+@FeignClient(name = "product-client-2", url = "http://localhost:8081/product/", configuration = FeignConfig.class)
+public interface ProductClient {
+
+    @RequestMapping(value = "{id}", method = RequestMethod.GET, produces = "application/json")
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClient.java:14]: Is this correct? Does the `getProduct` produce `application/json`? or does it `consume`?

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import feign.Logger;
+import feign.codec.ErrorDecoder;
+import org.springframework.context.annotation.Bean;
+
+public class FeignConfig {
+
+    @Bean
+    Logger.Level feignLoggerLevel() {
+		return Logger.Level.FULL;
+	}
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java:12]: The spacing's gone all funny here

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package com.baeldung.cloud.openfeign.customizederrorhandling.config;
+
+import feign.Logger;
+import feign.codec.ErrorDecoder;
+import org.springframework.context.annotation.Bean;
+
+public class FeignConfig {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java:7]: Should this have `@Configuration`?

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/client/ProductClient.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.springframework.cloud.openfeign.FeignClient;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+@FeignClient(name = "product-client", url = "http://localhost:8081/product/", configuration = FeignConfig.class)
+public interface ProductClient {
+
+    @RequestMapping(value = "{id}", method = RequestMethod.GET, produces = "application/json")
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/client/ProductClient.java:14]: `produces`? or `consumes`?

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/config/FeignConfig.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.baeldung.cloud.openfeign.defaulterrorhandling.config;
+
+import feign.Logger;
+import org.springframework.context.annotation.Bean;
+
+public class FeignConfig {
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/config/FeignConfig.java:6]: `@Configuration` ?

File: spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClientUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        stubFor(get(urlEqualTo("/product/" + productId))
+                .willReturn(aResponse().withStatus(404)));
+
+        Assert.assertThrows(ProductNotFoundException.class, () -> productClient.getProduct(productId));
+    }
+
+    @After
+    public void stopWireMockServer() {
+        wireMockServer.stop();
+    }
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClientUnitTest.java:57]: My preference is to move `@Before` and `@After` to the very top of the test fixture - this seems to be the standard across multiple languages/test frameworks.

File: spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClientUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Test
+    public void givenProductNotFound_whenGetProductCalled_thenThrowBadRequestException() {
+        String productId = "test";
+
+        stubFor(get(urlEqualTo("/product/" + productId))
+                .willReturn(aResponse().withStatus(404)));
+
+        Assert.assertThrows(ProductNotFoundException.class, () -> productClient.getProduct(productId));
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClientUnitTest.java:51]: Static import `assertThrows` for brevity

File: spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+        ErrorResponse expectedError = new ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR,
+        "Product Api is unavailable","uri=/myapp2/product/" + productId);
+
+        MvcResult result = mockMvc.perform(get("/myapp2/product/" + productId))
+            .andExpect(status().isInternalServerError()).andReturn();
+
+        ErrorResponse errorResponse = objectMapper.readValue(result.getResponse().getContentAsString(), ErrorResponse.class);
+
+        Assert.assertEquals(expectedError.getCode(), errorResponse.getCode());
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java:66]: static import all assertions for brevity

**Review by ashleyfrieze — CHANGES_REQUESTED** [review]

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClient.java (hunk 1 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClient.java]: I think it's correct as the Rest Api produces the resource as application/json and earlier article also mentioned the same thing in feign client. https://www.baeldung.com/spring-cloud-openfeign. Still, there is some element of doubt so i have removed that part as it works even without that.

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java (hunk 2 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java]: Ya funny :). I guess side effect of reusing the code already present in the project.

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java (hunk 3 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/FeignConfig.java:7]: It's not required as the class is passed to feign. And, this Spring configuration is not needed for that.

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/client/ProductClient.java (hunk 4 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/client/ProductClient.java]: Removed that for now. Works without that as well.

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/config/FeignConfig.java (hunk 5 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/defaulterrorhandling/config/FeignConfig.java:6]: It's not required as the class is passed to feign. The Spring configuration is not needed for that.

File: spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClientUnitTest.java (hunk 7 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/client/ProductClientUnitTest.java]: Added static import

File: spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java (hunk 8 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java]: Added static import

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/exception/ProductExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class ProductExceptionHandler extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler({ProductServiceNotAvailableException.class})
+    public ResponseEntity<ErrorResponse> handleProductServiceNotAvailableException(ProductServiceNotAvailableException exception, WebRequest request) {
+        return new ResponseEntity<>(new ErrorResponse(
+                HttpStatus.INTERNAL_SERVER_ERROR,
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/exception/ProductExceptionHandler.java:17]: The continuation indent should be two spaces

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/exception/ProductServiceNotAvailableException.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+public class ProductServiceNotAvailableException extends RuntimeException {
+
+    public ProductServiceNotAvailableException(String message) {
+        super(message);
+    }
+
+    @Override
+    public String toString() {
+        return getMessage();
+    }
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/exception/ProductServiceNotAvailableException.java:12]: Do we need this `toString`?

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/exception/ProductExceptionHandler.java (hunk 9 above)
Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/exception/ProductExceptionHandler.java]: changed to two spaces

File: spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    @Test
+    public void givenProductApiIsNotAvailable_whenGetProductCalled_ThenReturnInternalServerError() throws Exception {
+        String productId = "test";
+
+        stubFor(WireMock.get(urlEqualTo("/product/" + productId))
+            .willReturn(aResponse()
+            .withStatus(HttpStatus.SERVICE_UNAVAILABLE.value())));
+
+        ErrorResponse expectedError = new ErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR,
+        "Product Api is unavailable","uri=/myapp2/product/" + productId);
```
Comment (ashleyfrieze) [inline spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java:64]: continuation lines need 2 spaces of indent

Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/test/java/com/baeldung/cloud/openfeign/customizederrorhandling/controller/ProductControllerUnitTest.java]: included 2 spaces of indent.

File: spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/CustomErrorDecoder.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+import com.baeldung.cloud.openfeign.customizederrorhandling.exception.ProductNotFoundException;
+import com.baeldung.cloud.openfeign.customizederrorhandling.exception.ProductServiceNotAvailableException;
+import feign.Response;
+import feign.codec.ErrorDecoder;
+
+public class CustomErrorDecoder implements ErrorDecoder {
+
+    @Override
+    public Exception decode(String methodKey, Response response) {
+
```
Comment (KevinGilmore) [inline spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/CustomErrorDecoder.java:13]: Remove this blank line at the beginning of the method. In general, when we have any set of curly braces within methods, we should not have empty opening or closing lines.

Comment (saikatcse03) [reply spring-cloud-modules/spring-cloud-openfeign/src/main/java/com/baeldung/cloud/openfeign/customizederrorhandling/config/CustomErrorDecoder.java]: I have made this correction.

**Pull request merged**